File: packages/sdk/java/packages/android/src/main/java/dev/edgebase/sdk/client/DatabaseLiveClient.java
// EdgeBase Java SDK — Database live transport.
// OkHttp WebSocket-based DB subscription transport with auto-reconnect and message-based auth.
// Supports table subscriptions and server-side filters.
package dev.edgebase.sdk.client;

import dev.edgebase.sdk.core.*;
import dev.edgebase.sdk.core.generated.GeneratedDbApi;

import com.google.gson.Gson;
import okhttp3.*;

import java.io.Closeable;
import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.Consumer;

/**
 * DatabaseLive client using OkHttp WebSocket with automatic reconnection
 * and message-based authentication.
 *
 * <p>The server (DatabaseLiveDO) ignores HTTP Authorization headers for WebSocket
 * connections. Authentication is performed by sending a {@code {"type":"auth"}}
 * message immediately after the WebSocket opens.
 *
 * <p>
 * Usage:
 *
 * <pre>{@code
 * // Table subscription
 * var sub = client.db("shared").table("posts").onSnapshot(change -> System.out.println(change));
 * sub.close(); // unsubscribe
 *
 * // Collection subscription
 * var sub = client.db("shared").table("posts").onSnapshot(change -> System.out.println(change));
 * sub.close();
 * }</pre>
 */
class DatabaseLiveClient implements dev.edgebase.sdk.core.DatabaseLiveClient {
    private static final Gson gson = new Gson();
    private static final String SDK_VERSION = "0.2.5";

    private final String url;
    private final TokenManager tokenManager;
    private final OkHttpClient okHttpClient;

    private WebSocket webSocket;
    private volatile boolean isConnected;
    private volatile boolean authenticated;
    private volatile boolean waitingForAuth;
    private volatile boolean shouldReconnect = true;
    private volatile int reconnectAttempts;
    private final List<Consumer<Map<String, Object>>> messageListeners = new CopyOnWriteArrayList<>();
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
        Thread t = new Thread(r, "edgebase-dblive");
        t.setDaemon(true);
        return t;
    });

    /** Channels the client is currently subscribed to. */
    private final Set<String> subscribedChannels = ConcurrentHashMap.newKeySet();

    /** Server-side filters per channel for recovery after FILTER_RESYNC. */
    private final ConcurrentHashMap<String, List<FilterTuple>> channelFilters = new ConcurrentHashMap<>();

    /** Server-side OR filters per channel for recovery after FILTER_RESYNC. */
    private final ConcurrentHashMap<String, List<FilterTuple>> channelOrFilters = new ConcurrentHashMap<>();

    DatabaseLiveClient(String url, TokenManager tokenManager) {
        this.url = url;
        this.tokenManager = tokenManager;
        this.okHttpClient = new OkHttpClient.Builder()
                .readTimeout(0, TimeUnit.MILLISECONDS)
                .build();
        this.tokenManager.setOnAuthStateChange(this::handleAuthStateChange);
    }

    private static String normalizeDatabaseLiveChannel(String tableOrChannel) {
        return tableOrChannel.startsWith("dblive:") ? tableOrChannel : "dblive:" + tableOrChannel;
    }

    private static boolean matchesDatabaseLiveChannel(String channel, DbChange change, String messageChannel) {
        if (messageChannel != null && !messageChannel.isBlank()) {
            return Objects.equals(channel, normalizeDatabaseLiveChannel(messageChannel));
        }
        String[] parts = channel.split(":");
        if (parts.length == 0 || !"dblive".equals(parts[0])) {
            return false;
        }
        return switch (parts.length) {
            case 2 -> Objects.equals(parts[1], change.getTable());
            case 3 -> Objects.equals(parts[2], change.getTable());
            case 4 -> (Objects.equals(parts[2], change.getTable()) && Objects.equals(parts[3], change.getId()))
                    || Objects.equals(parts[3], change.getTable());
            default -> Objects.equals(parts[3], change.getTable()) && Objects.equals(parts[4], change.getId());
        };
    }

    private static String channelTableName(String channel) {
        String[] parts = channel.split(":");
        if (parts.length <= 1) {
            return channel;
        }
        if (parts.length == 2) {
            return parts[1];
        }
        if (parts.length == 3) {
            return parts[2];
        }
        return parts[3];
    }

    // ─── Connection ───

    private synchronized void ensureConnected(String channel) {
        if (isConnected)
            return;
        connect(channel);
    }

    private void connect(String channel) {
        String wsBase = url.replace("https://", "wss://").replace("http://", "ws://")
                + GeneratedDbApi.ApiPaths.CONNECT_DATABASE_SUBSCRIPTION;
        String wsUrl = channel != null
                ? wsBase + "?channel=" + URLEncoder.encode(channel, StandardCharsets.UTF_8)
                : wsBase;

        // No Authorization header — server ignores HTTP headers for WS auth.
        // Auth is performed via message after connection opens.
        Request request = new Request.Builder().url(wsUrl).build();

        webSocket = okHttpClient.newWebSocket(request, new WebSocketListener() {
            @Override
            public void onOpen(WebSocket ws, Response response) {
                isConnected = true;
                authenticated = false;
                // Send auth message immediately (message-based auth)
                sendAuthMessage();
            }

            @Override
            @SuppressWarnings("unchecked")
            public void onMessage(WebSocket ws, String text) {
                try {
                    Map<String, Object> msg = gson.fromJson(text, Map.class);
                    if (msg == null)
                        return;

                    String type = (String) msg.get("type");

                    // ── Auth success: mark authenticated & re-subscribe all channels
                    if ("auth_success".equals(type)) {
                        authenticated = true;
                        reconnectAttempts = 0;
                        resubscribeAll();
                        return;
                    }

                    // ── Auth refreshed: handle revoked channels, then re-subscribe
                    if ("auth_refreshed".equals(type)) {
                        authenticated = true;
                        reconnectAttempts = 0;
                        handleAuthRefreshed(msg);
                        resubscribeAll();
                        return;
                    }

                    // ── FILTER_RESYNC: server lost filter state (e.g. after hibernation)
                    if ("FILTER_RESYNC".equals(type)) {
                        resyncFilters();
                        return;
                    }

                    // Handle NOT_AUTHENTICATED: trigger re-auth
                    if ("error".equals(type)) {
                        String code = (String) msg.get("code");
                        if ("NOT_AUTHENTICATED".equals(code) || "AUTH_FAILED".equals(code)) {
                            handleAuthenticationFailure(new EdgeBaseError(401, "Authentication lost"));
                        }
                    }

                    // Dispatch to all registered listeners
                    for (Consumer<Map<String, Object>> listener : messageListeners) {
                        listener.accept(msg);
                    }
                } catch (Exception ignored) {
                }
            }

            @Override
            public void onClosing(WebSocket ws, int code, String reason) {
                ws.close(1000, null);
                isConnected = false;
                authenticated = false;
            }

            @Override
            public void onClosed(WebSocket ws, int code, String reason) {
                isConnected = false;
                authenticated = false;
                if (shouldReconnect && !waitingForAuth) {
                    long delay = Math.min((long) (1000 * Math.pow(2, reconnectAttempts)), 30000);
                    reconnectAttempts++;
                    scheduler.schedule(() -> connect(null), delay, TimeUnit.MILLISECONDS);
                }
            }

            @Override
            public void onFailure(WebSocket ws, Throwable t, Response response) {
                isConnected = false;
                authenticated = false;
                if (shouldReconnect && !waitingForAuth) {
                    long delay = Math.min((long) (1000 * Math.pow(2, reconnectAttempts)), 30000);
                    reconnectAttempts++;
                    scheduler.schedule(() -> connect(null), delay, TimeUnit.MILLISECONDS);
                }
            }
        });
    }

    // ─── Auth ───

    /**
     * Send a {@code {"type":"auth","token":"...","sdkVersion":"0.2.5"}} message.
     * This is the only auth mechanism the server accepts for WebSocket connections.
     */
    private void sendAuthMessage() {
        String token = tokenManager.getAccessToken();
        if (token == null) {
            handleAuthenticationFailure(new EdgeBaseError(401, hasSession()
                    ? "DatabaseLive is waiting for an active access token."
                    : "No access token available. Sign in first."));
            return;
        }

        Map<String, Object> authMsg = new LinkedHashMap<>();
        authMsg.put("type", "auth");
        authMsg.put("token", token);
        authMsg.put("sdkVersion", SDK_VERSION);
        sendRaw(authMsg);
    }

    /**
     * Handle {@code auth_refreshed} message with {@code revokedChannels}.
     * Removes revoked channels from subscription tracking and filter maps,
     * then dispatches {@code subscription_revoked} events so the application can react.
     */
    @SuppressWarnings("unchecked")
    private void handleAuthRefreshed(Map<String, Object> msg) {
        List<String> revoked = Collections.emptyList();
        Object revokedObj = msg.get("revokedChannels");
        if (revokedObj instanceof List) {
            revoked = (List<String>) revokedObj;
        }

        for (String channel : revoked) {
            String normalized = normalizeDatabaseLiveChannel(channel);
            subscribedChannels.remove(normalized);
            channelFilters.remove(normalized);
            channelOrFilters.remove(normalized);
        }

        // Dispatch subscription_revoked events to listeners
        if (!revoked.isEmpty()) {
            for (String channel : revoked) {
                Map<String, Object> event = new LinkedHashMap<>();
                event.put("type", "subscription_revoked");
                event.put("channel", channel);
                for (Consumer<Map<String, Object>> listener : messageListeners) {
                    listener.accept(event);
                }
            }
        }
    }

    // ─── Resubscribe / Resync ───

    /**
     * Re-subscribe all tracked channels after authentication.
     * Sends subscribe messages (with stored filters) for every channel
     * in {@link #subscribedChannels}.
     */
    private void resubscribeAll() {
        for (String channel : subscribedChannels) {
            sendSubscribe(channel);
        }
    }

    /**
     * Re-send stored filters to server after FILTER_RESYNC.
     * Called when the server signals it lost filter state (e.g. after hibernation).
     */
    private void resyncFilters() {
        for (String channel : subscribedChannels) {
            List<FilterTuple> filters = channelFilters.get(channel);
            List<FilterTuple> orFilters = channelOrFilters.get(channel);
            boolean hasFilters = filters != null && !filters.isEmpty();
            boolean hasOrFilters = orFilters != null && !orFilters.isEmpty();
            if (hasFilters || hasOrFilters) {
                Map<String, Object> msg = new LinkedHashMap<>();
                msg.put("type", "subscribe");
                msg.put("channel", channel);
                if (hasFilters) {
                    msg.put("filters", toJsonFilters(filters));
                }
                if (hasOrFilters) {
                    msg.put("orFilters", toJsonFilters(orFilters));
                }
                sendRaw(msg);
            }
        }
    }

    // ─── Send ───

    /**
     * Send a message through the WebSocket. Only sends if connected and authenticated.
     * Use {@link #sendRaw(Map)} for messages that bypass the auth check (e.g. the auth
     * message itself).
     */
    void sendMessage(Map<String, Object> message) {
        if (webSocket != null && isConnected && authenticated) {
            webSocket.send(gson.toJson(message));
        }
    }

    /**
     * Send a raw message without auth check. Used for the auth handshake itself.
     */
    private void sendRaw(Map<String, Object> message) {
        if (webSocket != null && isConnected) {
            webSocket.send(gson.toJson(message));
        }
    }

    /**
     * Send a subscribe message for the given channel, including stored filters if any.
     */
    private void sendSubscribe(String channel) {
        Map<String, Object> msg = new LinkedHashMap<>();
        msg.put("type", "subscribe");
        msg.put("channel", channel);

        List<FilterTuple> filters = channelFilters.get(channel);
        List<FilterTuple> orFilters = channelOrFilters.get(channel);
        if (filters != null && !filters.isEmpty()) {
            msg.put("filters", toJsonFilters(filters));
        }
        if (orFilters != null && !orFilters.isEmpty()) {
            msg.put("orFilters", toJsonFilters(orFilters));
        }

        sendRaw(msg);
    }

    // ─── Subscribe ───

    /**
     * Subscribe to a table and receive changes.
     *
     * @param tableName table name
     * @param listener  callback for each change
     * @return a Subscription that can be closed to unsubscribe
     */
    @Override
    public Subscription subscribe(String tableName, Consumer<DbChange> listener) {
        return subscribe(tableName, listener, null, null);
    }

    /**
     * Subscribe to a table with server-side filters and receive changes.
     *
     * @param tableName     table name
     * @param listener      callback for each change
     * @param serverFilters server-side filter conditions, may be null
     * @param serverOrFilters server-side OR filter conditions, may be null
     * @return a Subscription that can be closed to unsubscribe
     */
    public Subscription subscribe(String tableName, Consumer<DbChange> listener,
            List<FilterTuple> serverFilters, List<FilterTuple> serverOrFilters) {
        String channel = normalizeDatabaseLiveChannel(tableName);

        // Track channel
        subscribedChannels.add(channel);

        // Store server-side filters for recovery
        if (serverFilters != null && !serverFilters.isEmpty()) {
            channelFilters.put(channel, new ArrayList<>(serverFilters));
        }
        if (serverOrFilters != null && !serverOrFilters.isEmpty()) {
            channelOrFilters.put(channel, new ArrayList<>(serverOrFilters));
        }

        ensureConnected(channel);

        // Only send subscribe if already authenticated; otherwise resubscribeAll() will
        // handle it once auth_success arrives.
        if (authenticated) {
            sendSubscribe(channel);
        }

        Consumer<Map<String, Object>> messageListener = msg -> {
            String type = (String) msg.get("type");
            if ("db_change".equals(type)) {
                DbChange change = DbChange.fromJson(msg);
                String messageChannel = msg.get("channel") instanceof String ? (String) msg.get("channel") : null;
                if (matchesDatabaseLiveChannel(channel, change, messageChannel)) {
                    listener.accept(change);
                }
            } else if ("batch_changes".equals(type)) {
                Object rawChanges = msg.get("changes");
                if (!(rawChanges instanceof List<?> changes)) {
                    return;
                }
                String fallbackTable = msg.get("table") instanceof String
                        ? (String) msg.get("table")
                        : channelTableName(msg.get("channel") instanceof String ? (String) msg.get("channel") : "");
                for (Object item : changes) {
                    if (!(item instanceof Map<?, ?> changeMap)) {
                        continue;
                    }
                    Map<String, Object> synthetic = new LinkedHashMap<>();
                    synthetic.put("changeType", changeMap.get("event"));
                    synthetic.put("table", fallbackTable);
                    synthetic.put("docId", changeMap.get("docId"));
                    synthetic.put("data", changeMap.get("data"));
                    synthetic.put("timestamp", changeMap.get("timestamp"));
                    DbChange change = DbChange.fromJson(synthetic);
                    String messageChannel = msg.get("channel") instanceof String ? (String) msg.get("channel") : null;
                    if (matchesDatabaseLiveChannel(channel, change, messageChannel)) {
                        listener.accept(change);
                    }
                }
            }
        };
        messageListeners.add(messageListener);

        return new Subscription(channel, messageListener);
    }

    @Override
    public void unsubscribe(String id) {
        String channel = normalizeDatabaseLiveChannel(id);
        subscribedChannels.remove(channel);
        channelFilters.remove(channel);
        channelOrFilters.remove(channel);
        if (authenticated) {
            sendRaw(Map.of(
                    "type", "unsubscribe",
                    "channel", channel));
        }
    }

    // ─── Cleanup ───

    void destroy() {
        shouldReconnect = false;
        if (webSocket != null) {
            webSocket.close(1000, "Client destroyed");
            webSocket = null;
        }
        isConnected = false;
        authenticated = false;
        messageListeners.clear();
        subscribedChannels.clear();
        channelFilters.clear();
        channelOrFilters.clear();
        scheduler.shutdownNow();
        okHttpClient.dispatcher().executorService().shutdown();
        okHttpClient.connectionPool().evictAll();
        Cache cache = okHttpClient.cache();
        if (cache != null) {
            try {
                cache.close();
            } catch (IOException ignored) {
            }
        }
    }

    // ─── Helpers ───

    /**
     * Convert a list of {@link FilterTuple} to the JSON-serializable format
     * expected by the server: {@code [[field, op, value], ...]}.
     */
    private static List<List<Object>> toJsonFilters(List<FilterTuple> filters) {
        List<List<Object>> result = new ArrayList<>();
        for (FilterTuple f : filters) {
            result.add(f.toJson());
        }
        return result;
    }

    // ─── Subscription ───

    /**
     * A database-live subscription that can be closed to unsubscribe.
     */
    public class Subscription implements Closeable, dev.edgebase.sdk.core.DatabaseLiveClient.Subscription {
        private final String tableName;
        private final Consumer<Map<String, Object>> messageListener;
        private volatile boolean closed;

        Subscription(String tableName, Consumer<Map<String, Object>> messageListener) {
            this.tableName = tableName;
            this.messageListener = messageListener;
        }

        @Override
        public void close() {
            if (closed)
                return;
            closed = true;
            messageListeners.remove(messageListener);
            String channel = normalizeDatabaseLiveChannel(tableName);
            subscribedChannels.remove(channel);
            channelFilters.remove(channel);
            channelOrFilters.remove(channel);
            if (authenticated) {
                sendRaw(Map.of(
                        "type", "unsubscribe",
                        "channel", channel));
            }
        }

        @Override
        public void cancel() {
            close();
        }

        public boolean isClosed() {
            return closed;
        }
    }

    private void refreshAuth() {
        String token = tokenManager.getAccessToken();
        if (token == null || !isConnected) {
            return;
        }
        sendRaw(Map.of(
                "type", "auth",
                "token", token,
                "sdkVersion", SDK_VERSION));
    }

    private void handleAuthStateChange(Map<String, Object> user) {
        if (user != null) {
            if (isConnected && authenticated) {
                refreshAuth();
                return;
            }
            waitingForAuth = false;
            if (!subscribedChannels.isEmpty() && !isConnected) {
                connect(subscribedChannels.iterator().next());
            }
            return;
        }

        waitingForAuth = !subscribedChannels.isEmpty();
        isConnected = false;
        authenticated = false;
        if (webSocket != null) {
            webSocket.close(1000, "Signed out");
            webSocket = null;
        }
    }

    private void handleAuthenticationFailure(EdgeBaseError error) {
        waitingForAuth = error.getStatusCode() == 401 && !subscribedChannels.isEmpty() && !hasSession();
        isConnected = false;
        authenticated = false;
        if (webSocket != null) {
            webSocket.close(4001, error.getMessage());
            webSocket = null;
        }
    }

    private boolean hasSession() {
        return tokenManager.getRefreshToken() != null || tokenManager.currentUser() != null;
    }
}

File: packages/sdk/kotlin/client/src/commonMain/kotlin/dev/edgebase/sdk/client/DatabaseLiveClient.kt
// EdgeBase Kotlin SDK — Database live client (KMP).
//
// Ktor WebSocket-based subscriptions with auto-reconnect.
// Supports table subscriptions (Flow) and server-side filters.
// Auth via WebSocket message (not HTTP headers) — server database live endpoint requires
// a { type: "auth", token, sdkVersion } message after session open.
//: OkHttp WebSocket → Ktor WebSocket for KMP.

package dev.edgebase.sdk.client

import dev.edgebase.sdk.core.*
import dev.edgebase.sdk.core.DatabaseLiveClient as CoreDatabaseLiveClient
import dev.edgebase.sdk.core.generated.ApiPaths

import io.ktor.client.plugins.websocket.*
import io.ktor.client.request.*
import io.ktor.websocket.*

import kotlinx.coroutines.*
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.*
import kotlinx.serialization.json.*

private const val SDK_VERSION = "0.1.5"

// MARK: - FilterTuple

/**
 * Server-side filter condition: [field, operator, value].
 * Mirrors the JS SDK FilterTuple type.
 */
typealias FilterTuple = Triple<String, String, Any?>

private fun normalizeDatabaseLiveChannel(tableOrChannel: String): String =
    if (tableOrChannel.startsWith("dblive:")) tableOrChannel else "dblive:$tableOrChannel"

private fun channelTableName(channel: String): String {
    val parts = channel.split(":")
    return when {
        parts.size <= 1 -> channel
        parts.size == 2 -> parts[1]
        parts.size == 3 -> parts[2]
        else -> parts[3]
    }
}

private fun matchesDatabaseLiveChannel(channel: String, change: DbChange, messageChannel: String? = null): Boolean {
    if (!messageChannel.isNullOrBlank()) {
        return channel == normalizeDatabaseLiveChannel(messageChannel)
    }
    val parts = channel.split(":")
    if (parts.firstOrNull() != "dblive") return false
    return when (parts.size) {
        2 -> parts[1] == change.table
        3 -> parts[2] == change.table
        4 -> {
            // Could be dblive:ns:table:docId or dblive:ns:instanceId:table
            if (parts[2] == change.table && change.id == parts[3]) true
            else parts[3] == change.table
        }
        else -> parts[3] == change.table && change.id == parts[4]
    }
}

// MARK: - DatabaseLiveClient

/**
 * Database live client using Ktor WebSocket with automatic reconnection.
 *
 * Authentication is performed via WebSocket message (not HTTP headers):
 * after the WS session opens, we send `{"type":"auth","token":"...","sdkVersion":"0.1.5"}`
 * and wait for `auth_success` or `auth_refreshed` before signaling ready.
 */
internal class DatabaseLiveClient(
    private val url: String,
    private val tokenManager: TokenManager
) : CoreDatabaseLiveClient {
    private val json = Json { ignoreUnknownKeys = true }
    private val ktorClient = createPlatformHttpClient().config {
        install(WebSockets)
    }
    private var session: DefaultClientWebSocketSession? = null
    private var isConnected = false
    private var isAuthenticated = false
    private var shouldReconnect = true
    private var connectionJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    // Shared message flow for subscription lifecycle events.
    internal val messageFlow = MutableSharedFlow<Map<String, Any?>>(
        extraBufferCapacity = 64,
        onBufferOverflow = BufferOverflow.SUSPEND
    )

    /** Server-side filters per channel for recovery after FILTER_RESYNC. */
    private val channelFilters = mutableMapOf<String, List<FilterTuple>>()

    /** Server-side OR filters per channel for recovery after FILTER_RESYNC. */
    private val channelOrFilters = mutableMapOf<String, List<FilterTuple>>()

    /** Channels currently subscribed to — used for resubscribeAll() after auth. */
    private val subscribedChannels = mutableSetOf<String>()
    private var waitingForAuth = false

    init {
        if (tokenManager is ClientTokenManager) {
            tokenManager.setOnAuthStateChange { user ->
                handleAuthStateChange(user)
            }
        }
    }

    // MARK: - Connection

    private suspend fun ensureConnected(channel: String? = null) {
        if (isConnected && isAuthenticated) return
        val deferred = CompletableDeferred<Unit>()
        connect(channel, deferred)
        deferred.await()
    }

    private fun connect(channel: String? = null, readySignal: CompletableDeferred<Unit>? = null) {
        connectionJob?.cancel()
        connectionJob = scope.launch {
            try {
                val wsBase = url
                    .replace("https://", "wss://")
                    .replace("http://", "ws://") + ApiPaths.CONNECT_DATABASE_SUBSCRIPTION
                val wsUrl = if (channel != null) {
                    "$wsBase?channel=${platformUrlEncode(channel)}"
                } else wsBase

                ktorClient.webSocket(wsUrl) {
                    session = this
                    isConnected = true

                    // --- WebSocket auth handshake ---
                    // Server DatabaseLiveDO ignores HTTP headers; auth must be sent as a
                    // WebSocket message and we must wait for auth_success/auth_refreshed.
                    try {
                        authenticate(readySignal)
                    } catch (e: Exception) {
                        handleAuthenticationFailure(e)
                        readySignal?.complete(Unit) // Don't hang on auth failure
                        return@webSocket
                    }

                    // --- Normal incoming frame loop (post-auth) ---
                    try {
                        for (frame in incoming) {
                            if (frame is Frame.Text) {
                                processIncomingFrame(frame)
                            }
                        }
                    } catch (_: Exception) { /* connection closed */ }
                }
            } catch (_: Exception) {
                readySignal?.complete(Unit) // Don't hang on connection failure
            }

            isConnected = false
            isAuthenticated = false
            session = null

            if (shouldReconnect && !waitingForAuth) {
                delay(1000)
                connect()
            }
        }
    }

    // MARK: - Auth Handshake

    /**
     * Send auth message and wait for auth_success or auth_refreshed.
     *
     * This runs inside the `ktorClient.webSocket()` block, reading frames
     * from the incoming channel until the auth handshake completes.
     */
    private suspend fun DefaultClientWebSocketSession.authenticate(
        readySignal: CompletableDeferred<Unit>?
    ) {
        val token = tokenManager.getAccessToken()
        if (token == null) {
            val hasSession = tokenManager.getRefreshToken() != null
            val message = if (hasSession) {
                "DatabaseLive is waiting for an active access token."
            } else {
                "No access token available. Sign in first."
            }
            throw EdgeBaseError(401, message)
        }

        // Send auth message via the WebSocket session
        val authMsg = buildJsonObject {
            put("type", JsonPrimitive("auth"))
            put("token", JsonPrimitive(token))
            put("sdkVersion", JsonPrimitive(SDK_VERSION))
        }
        send(Frame.Text(json.encodeToString(JsonElement.serializer(), authMsg)))

        // Read frames until we get auth_success, auth_refreshed, or error
        for (frame in incoming) {
            if (frame !is Frame.Text) continue

            val text = frame.readText()
            val element = try {
                json.parseToJsonElement(text)
            } catch (_: Exception) { continue }

            @Suppress("UNCHECKED_CAST")
            val msg = HttpClient.jsonElementToAny(element) as? Map<String, Any?> ?: continue
            val type = msg["type"] as? String

            when (type) {
                "auth_success" -> {
                    isAuthenticated = true
                    resubscribeAll()
                    readySignal?.complete(Unit)
                    return
                }
                "auth_refreshed" -> {
                    handleAuthRefreshed(msg)
                    isAuthenticated = true
                    resubscribeAll()
                    readySignal?.complete(Unit)
                    return
                }
                "error" -> {
                    val errorMsg = msg["message"] as? String ?: "Auth failed"
                    throw EdgeBaseError(401, errorMsg)
                }
            }
            // Ignore unrelated messages during auth handshake
        }
        // If incoming channel closes without auth response, throw
        throw EdgeBaseError(401, "WebSocket closed before auth completed")
    }

    // MARK: - Incoming Frame Processing

    /**
     * Process an incoming text frame after authentication.
     * Intercepts auth_refreshed and FILTER_RESYNC before emitting to messageFlow.
     */
    private fun processIncomingFrame(frame: Frame.Text) {
        try {
            val text = frame.readText()
            val element = json.parseToJsonElement(text)
            @Suppress("UNCHECKED_CAST")
            val msg = HttpClient.jsonElementToAny(element) as? Map<String, Any?> ?: return

            val type = msg["type"] as? String

            // Intercept auth_refreshed: clean up revoked channels and dispatch
            if (type == "auth_refreshed") {
                handleAuthRefreshed(msg)
                // Dispatch subscription_revoked events to messageFlow for app listeners
                @Suppress("UNCHECKED_CAST")
                val revoked = (msg["revokedChannels"] as? List<*>)
                    ?.filterIsInstance<String>() ?: emptyList()
                for (ch in revoked) {
                    messageFlow.tryEmit(mapOf(
                        "type" to "subscription_revoked",
                        "channel" to ch
                    ))
                }
                return
            }

            // Intercept FILTER_RESYNC: re-send stored filters to server
            if (type == "FILTER_RESYNC") {
                resyncFilters()
                return
            }

            if (type == "batch_changes") {
                @Suppress("UNCHECKED_CAST")
                val changes = msg["changes"] as? List<Map<String, Any?>> ?: emptyList()
                val table = (msg["table"] as? String)
                    ?: channelTableName(msg["channel"] as? String ?: "")
                for (change in changes) {
                    messageFlow.tryEmit(mapOf(
                        "type" to "db_change",
                        "changeType" to change["event"],
                        "table" to table,
                        "docId" to change["docId"],
                        "data" to change["data"],
                        "timestamp" to change["timestamp"],
                    ))
                }
                return
            }

            // Handle NOT_AUTHENTICATED: attempt re-auth
            if (type == "error") {
                val code = msg["code"] as? String
                if (code == "NOT_AUTHENTICATED") {
                    isAuthenticated = false
                    scope.launch {
                        try {
                            session?.close(CloseReason(CloseReason.Codes.NORMAL, "Re-authenticating"))
                        } catch (_: Exception) { /* already closed */ }
                        session = null
                        isConnected = false
                        connect()
                    }
                }
            }

            // All other messages: emit to shared flow
            messageFlow.tryEmit(msg)
        } catch (_: Exception) { /* ignore parse failures */ }
    }

    // MARK: - Auth Refreshed Handling

    /**
     * Handle auth_refreshed: parse revokedChannels and clean up state.
     */
    private fun handleAuthRefreshed(msg: Map<String, Any?>) {
        @Suppress("UNCHECKED_CAST")
        val revoked = (msg["revokedChannels"] as? List<*>)
            ?.filterIsInstance<String>() ?: emptyList()
        for (ch in revoked) {
            subscribedChannels.remove(ch)
            channelFilters.remove(ch)
            channelOrFilters.remove(ch)
        }
    }

    // MARK: - Resubscribe / Resync

    /**
     * Re-subscribe all tracked channels with stored filters after auth.
     */
    private fun resubscribeAll() {
        for (channel in subscribedChannels.toList()) {
            sendSubscribeInternal(channel)
        }
    }

    private fun refreshAuth() {
        scope.launch {
            val token = tokenManager.getAccessToken() ?: return@launch
            sendMessage(mapOf(
                "type" to "auth",
                "token" to token,
                "sdkVersion" to SDK_VERSION
            ))
        }
    }

    private fun handleAuthStateChange(user: Map<String, Any?>?) {
        if (user != null) {
            if (isConnected && isAuthenticated) {
                refreshAuth()
                return
            }

            waitingForAuth = false
            if (subscribedChannels.isNotEmpty() && !isConnected) {
                connect(subscribedChannels.firstOrNull())
            }
            return
        }

        waitingForAuth = subscribedChannels.isNotEmpty()
        isConnected = false
        isAuthenticated = false
        scope.launch {
            try {
                session?.close(CloseReason(CloseReason.Codes.NORMAL, "Signed out"))
            } catch (_: Exception) { /* already closed */ }
            session = null
        }
    }

    private fun handleAuthenticationFailure(error: Exception) {
        waitingForAuth = error is EdgeBaseError
            && error.statusCode == 401
            && subscribedChannels.isNotEmpty()
        isConnected = false
        isAuthenticated = false
        scope.launch {
            try {
                session?.close(CloseReason(CloseReason.Codes.VIOLATED_POLICY, error.message ?: "Auth failed"))
            } catch (_: Exception) { /* already closed */ }
            session = null
        }
    }

    /**
     * Re-send stored filters to server after FILTER_RESYNC.
     */
    private fun resyncFilters() {
        for (channel in channelFilters.keys.toList()) {
            val filters = channelFilters[channel] ?: emptyList()
            val orFilters = channelOrFilters[channel] ?: emptyList()
            if (filters.isNotEmpty() || orFilters.isNotEmpty()) {
                val msg = mutableMapOf<String, Any?>(
                    "type" to "subscribe",
                    "channel" to channel
                )
                if (filters.isNotEmpty()) msg["filters"] = filters.toFilterJson()
                if (orFilters.isNotEmpty()) msg["orFilters"] = orFilters.toFilterJson()
                sendMessage(msg)
            }
        }
    }

    // MARK: - Send

    internal fun sendMessage(message: Map<String, Any?>) {
        val element = HttpClient.anyToJsonElement(message)
        val str = json.encodeToString(JsonElement.serializer(), element)
        scope.launch {
            try {
                session?.send(Frame.Text(str))
            } catch (_: Exception) { /* ignore if disconnected */ }
        }
    }

    /**
     * Send a subscribe message for a channel, including stored filters.
     */
    private fun sendSubscribeInternal(channel: String) {
        val msg = mutableMapOf<String, Any?>(
            "type" to "subscribe",
            "channel" to channel
        )
        val filters = channelFilters[channel]
        val orFilters = channelOrFilters[channel]
        if (filters != null && filters.isNotEmpty()) msg["filters"] = filters.toFilterJson()
        if (orFilters != null && orFilters.isNotEmpty()) msg["orFilters"] = orFilters.toFilterJson()
        sendMessage(msg)
    }

    // MARK: - Subscribe

    /**
     * Subscribe to a table and receive changes as a Flow.
     */
    override fun subscribe(tableName: String): Flow<DbChange> = subscribe(
        tableName = tableName,
        serverFilters = null,
        serverOrFilters = null
    )

    /**
     * Subscribe to a table with optional server-side filters and receive changes as a Flow.
     */
    fun subscribe(
        tableName: String,
        serverFilters: List<FilterTuple>? = null,
        serverOrFilters: List<FilterTuple>? = null
    ): Flow<DbChange> = callbackFlow {
        val channel = normalizeDatabaseLiveChannel(tableName)

        // Store filters for recovery
        if (serverFilters != null && serverFilters.isNotEmpty()) {
            channelFilters[channel] = serverFilters
        }
        if (serverOrFilters != null && serverOrFilters.isNotEmpty()) {
            channelOrFilters[channel] = serverOrFilters
        }

        // Track channel subscription
        subscribedChannels.add(channel)

        ensureConnected(channel)

        // Send subscribe message with filters
        sendSubscribeInternal(channel)

        val job = scope.launch {
            messageFlow.collect { msg ->
                val type = msg["type"] as? String
                if (type == "db_change") {
                    val change = DbChange.fromJson(msg)
                    val messageChannel = msg["channel"] as? String
                    if (matchesDatabaseLiveChannel(channel, change, messageChannel)) {
                        trySend(change)
                    }
                }
            }
        }

        awaitClose {
            job.cancel()
            subscribedChannels.remove(channel)
            channelFilters.remove(channel)
            channelOrFilters.remove(channel)
            sendMessage(mapOf(
                "type" to "unsubscribe",
                "channel" to channel
            ))
        }
    }

    /**
     * Unsubscribe from a channel by ID.
     */
    override fun unsubscribe(id: String) {
        subscribedChannels.remove(id)
        channelFilters.remove(id)
        channelOrFilters.remove(id)
        sendMessage(mapOf("type" to "unsubscribe", "channel" to id))
    }

    // MARK: - Cleanup

    fun destroy() {
        shouldReconnect = false
        // Cancel connection job — this aborts the WebSocket session.
        // Server detects the disconnect automatically.
        connectionJob?.cancel()
        session = null
        isConnected = false
        isAuthenticated = false
        subscribedChannels.clear()
        channelFilters.clear()
        channelOrFilters.clear()
        scope.cancel()
        ktorClient.close()
    }
}

// MARK: - FilterTuple JSON Conversion

/**
 * Convert a list of FilterTuples to the JSON-compatible list-of-lists format
 * expected by the server: [[field, op, value], ...]
 */
private fun List<FilterTuple>.toFilterJson(): List<List<Any?>> =
    map { (field, op, value) -> listOf(field, op, value) }
